File: macos/Tests/ClearPowerCoreTests/GoldenTests.swift
// Golden tests: the Python daemon (daemon/clearpowerd) is the reference. Fixtures are
// produced by macos/scripts/gen-fixtures.py and compared value by value.
import Testing
import Foundation
@testable import ClearPowerCore

private func fixture(_ name: String) -> Any {
    let url = Bundle.module.url(forResource: name, withExtension: "json", subdirectory: "Fixtures")!
    return try! JSONSerialization.jsonObject(with: Data(contentsOf: url))
}

private func near(_ a: Double, _ b: Double, _ tol: Double = 1e-9, _ msg: String = "", sourceLocation: SourceLocation = #_sourceLocation) {
    let ok = abs(a - b) <= max(tol, abs(b) * 1e-9)
    #expect(ok, "\(msg): got \(a), expected \(b)", sourceLocation: sourceLocation)
}

struct EmaGoldenTests {
    @Test func ema() {
        for c in fixture("ema") as! [[String: Any]] {
            var e = Ema(tau: c["tau"] as! Double)
            let steps = c["steps"] as! [[String: Any]]
            for (i, s) in steps.enumerated() {
                if i == steps.count - 1 { e.reset() }
                let v = e.update(s["x"] as! Double, at: s["t"] as! Double)
                near(v, s["v"] as! Double, 1e-9, "tau \(c["tau"]!) step \(i)")
            }
        }
    }
}

struct RuntimeGoldenTests {
    @Test func runtime() {
        var rt = RuntimeEstimator()
        // Fixture only records every 7th step (plus the first 7); replay the same plan.
        var t = 0.0, e = 60.0
        let plan: [(String, Double, Int)] = [("Discharging", 12, 180), ("Charging", -30, 120), ("Not charging", 0, 12), ("Discharging", 8, 60)]
        let recorded = fixture("runtime") as! [[String: Any]]
        var ri = 0, i = 0
        for (status, w, n) in plan {
            for _ in 0..<n {
                t += 10; e -= w * 10 / 3600
                rt.add(t: t, energyWh: e, status: status)
                i += 1
                if i % 7 == 0 || i < 8 {
                    let r = recorded[ri]; ri += 1
                    near(t, r["t"] as! Double)
                    let out = rt.estimate(energyNowWh: e, targetWh: r["target_wh"] as! Double, fallbackW: r["fallback_w"] as! Double)
                    let exp = r["out"] as! [String: Any]
                    for (k, v) in exp {
                        near(Snapshot.double(out[k])!, Snapshot.double(v)!, 1e-6, "step \(i) key \(k)")
                    }
                }
            }
        }
        #expect(ri == recorded.count)
    }
}

struct PowerModelGoldenTests {
    @Test func breakdown() {
        var m = PowerModel(smoothingS: 5)
        for (i, c) in (fixture("power_model") as! [[String: Any]]).enumerated() {
            let inp = c["in"] as! [String: Any]
            let raw = RawPower(batW: inp["bat_w"] as! Double, psys: inp["psys"] as! Double, package: inp["package"] as! Double,
                               core: inp["core"] as! Double, uncore: inp["uncore"] as! Double, dram: inp["dram"] as! Double)
            let out = m.update(raw: raw, onAC: inp["on_ac"] as! Bool, now: c["t"] as! Double,
                               displayEmission: inp["emission"] as! Double, displayOn: inp["display_on"] as! Bool)
            near(m.raw.rest, c["raw_rest"] as! Double, 1e-9, "step \(i) rest")
            for (k, v) in c["out"] as! [String: Any] {
                if let s = v as? String { #expect(out[k] as? String == s, Comment(rawValue: "step \(i) \(k)")) }
                else { near(Snapshot.double(out[k])!, Snapshot.double(v)!, 1e-9, "step \(i) \(k)") }
            }
        }
    }
}

final class FakeChargeHW: ChargeHardware {
    var thresholdsSupported = true
    var behaviours = ["auto", "inhibit-charge", "force-discharge"]
    var writes: [[Any]] = []
    var saved: Int? = nil
    func writeThresholds(start: Int, end: Int) throws {
        // Linux writes them in an order dictated by sysfs; the fixture is normalised below.
        writes.append(["start", start]); writes.append(["end", end])
    }
    func writeBehaviour(_ behaviour: String) throws { writes.append(["beh", behaviour]) }
    func loadLimit() -> Int? { nil }
    func saveLimit(_ limit: Int) { saved = limit }
}

struct ChargeGoldenTests {
    @Test func stateMachine() throws {
        let hw = FakeChargeHW()
        let c = ChargeStateMachine(hardware: hw, dischargeFloorPct: 20)
        for step in fixture("charge") as! [[String: Any]] {
            let op = step["op"] as! String
            let args = step["args"] as! [String: Any]
            switch op {
            case "startup": c.applyStartup()
            case "set_limit": try c.setLimit(args["pct"] as! Int)
            case "start_topup": try c.startTopUp()
            case "start_discharge": try c.startDischarge(target: args["target"] as! Int)
            case "cancel": try c.cancel()
            case "shutdown": c.shutdown()
            case "tick": c.tick(batPct: args["pct"] as! Int, batStatus: args["status"] as! String)
            default: Issue.record("unknown op \(op)")
            }
            let st = step["state"] as! [String: Any]
            #expect(c.mode.rawValue == st["charge_mode"] as! String, Comment(rawValue: String(describing: op)))
            #expect(c.limit == st["charge_limit"] as! Int, Comment(rawValue: String(describing: op)))
            #expect(c.target == st["charge_target"] as! Int, Comment(rawValue: String(describing: op)))
            // Compare the set of writes ignoring the sysfs-specific threshold ordering.
            let expected = (step["writes"] as! [[Any]]).map { "\($0[0]):\($0[1])" }.sorted()
            let got = hw.writes.map { "\($0[0]):\($0[1])" }.sorted()
            #expect(got == expected, Comment(rawValue: String(describing: op)))
            hw.writes = []
        }
    }
}

final class FakeBrightness: BrightnessControl {
    var available = true
    var max = 1000
    var value = 700
    var sets: [Int] = []
    func readRaw() -> Int? { value }
    func setRaw(_ v: Int) throws { value = v; sets.append(v) }
}

struct DisplayCalGoldenTests {
    @Test func sweepAndInterpolation() {
        let fx = fixture("display_cal") as! [String: Any]
        let bl = FakeBrightness()
        let d = DisplayCalibration(brightness: bl, storage: nil)
        var now = 100.0
        d.start(now: now, energyOK: true)
        #expect(d.state == "running")
        var guardCount = 0
        while d.state == "running" && guardCount < 10000 {
            now += 0.5
            let lvl = Double(bl.value) / 1000.0
            let noise = [0.3, -0.2, 0.1, 0.0, -0.1][Int(now * 2) % 5]
            d.tick(restRawW: 3.0 + 6.0 * lvl + noise, now: now)
            guardCount += 1
        }
        #expect(d.state == fx["final_state"] as! String)
        #expect(bl.sets == fx["sets"] as! [Int])
        let table = fx["table"] as! [[Double]]
        #expect(d.table.count == table.count)
        for (p, q) in zip(d.table, table) {
            #expect(Double(p.raw) == q[0]); near(p.w, q[1], 1e-9)
        }
        near(d.rest0, fx["rest0"] as! Double, 1e-9)
        for c in fx["interp"] as! [[String: Any]] {
            let raw = c["raw"] as! Int
            let apl = c["apl"] as! Double
            if let w = c["w"] as? Double {
                if apl >= 0 { d.setContent(apl, now: now) } else { d.setContent(-1, now: now - 100) }
                near(d.emissionW(rawBrightness: raw, now: now), w, 1e-9, "raw \(raw) apl \(apl)")
            } else if let w = c["w_stale"] as? Double {
                d.setContent(apl, now: now)
                near(d.emissionW(rawBrightness: raw, now: now + 61), w, 1e-9, "stale")
            }
        }
    }
}

struct HistoryGoldenTests {
    @Test func downsampling() {
        let fx = fixture("history") as! [String: Any]
        var h = History(seconds: 120, stepS: 10)
        for s in fx["snaps"] as! [[String: Any]] { h.add(s) }
        let get = fx["get"] as! [String: Any]
        for f in ["sys_w", "soc_w", "bat_pct"] {
            let exp = get[f] as! [[Double]]
            let got = h.get(f, seconds: 60)
            #expect(got.count == exp.count, Comment(rawValue: String(describing: f)))
            for (a, b) in zip(got, exp) { near(a.0, b[0]); near(a.1, b[1], 1e-9, f) }
        }
        let all = get["all_sys"] as! [[Double]]
        #expect(h.get("sys_w", seconds: 1e9).count == all.count)
    }
}

struct I18nTests {
    @Test func resolveAndFormat() {
        #expect(I18n.resolveLanguage("system", systemLanguages: ["zh-Hans-CN", "en"]) == "zh_CN")
        #expect(I18n.resolveLanguage("system", systemLanguages: ["en-US"]) == "en")
        #expect(I18n.resolveLanguage("zh-cn", systemLanguages: ["en"]) == "zh_CN")
        I18n.setLanguage("en", systemLanguages: [])
        #expect(I18n.t("limit", ["n": 80]) == "Limit 80%")
        #expect(I18n.fmtDuration(minutes: 125) == "2 h 5 m")
        #expect(I18n.fmtDuration(minutes: 12.4) == "12 min")
        I18n.setLanguage("zh-cn", systemLanguages: [])
        #expect(I18n.fmtDuration(minutes: 125) == "2 小时 5 分")
        #expect(I18n.t("missingKey") == "missingKey")
        #expect(Set(I18n.strings["en"]!.keys) == Set(I18n.strings["zh_CN"]!.keys), Comment(rawValue: String(describing: "zh/en key sets differ")))
    }
}

struct ProcessBudgetTests {
    @Test func budgetDistribution() {
        var pb = ProcessBudget(intervalS: 3, floorWindowS: 600)
        _ = pb.sample(now: 0, packageW: 4, usage: { [] })
        let top = pb.sample(now: 10, packageW: 10, usage: { [("a", 50), ("b", 25), ("a", 25), ("c", 0), ("d", 5)] })
        // floor is min over window = 4, budget 6 W; a=75 of 105
        #expect(top.count == 3)
        #expect(top[0].name == "a"); near(top[0].w, 6 * 75.0 / 105.0, 1e-9)
        #expect(top[1].name == "b"); #expect(top[2].name == "d")
        // Called again within interval -> cached
        #expect(pb.sample(now: 11, packageW: 20, usage: { [("z", 1)] }).first?.name == "a")
    }
}

File: windows/Tests/ClearPowerCoreTests/GoldenTests.cs
// Golden tests: the Python daemon (daemon/clearpowerd) is the reference. Fixtures are
// produced by macos/scripts/gen-fixtures.py (shared with the Swift port) and compared
// value by value. Mirrors macos/Tests/ClearPowerCoreTests/GoldenTests.swift.
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using ClearPower.Core;
using Xunit;

namespace ClearPower.Core.Tests
{
    internal static class Fx
    {
        public static object? Load(string name)
        {
            var path = Path.Combine(AppContext.BaseDirectory, "Fixtures", name + ".json");
            return Json.Parse(File.ReadAllText(path));
        }

        public static List<Dictionary<string, object?>> List(string name)
            => ((List<object?>)Load(name)!).Cast<Dictionary<string, object?>>().ToList();

        public static Dictionary<string, object?> Obj(string name) => (Dictionary<string, object?>)Load(name)!;

        public static void Near(double a, double b, double tol = 1e-9, string msg = "")
        {
            var ok = Math.Abs(a - b) <= Math.Max(tol, Math.Abs(b) * 1e-9);
            Assert.True(ok, $"{msg}: got {a}, expected {b}");
        }

        public static List<Dictionary<string, object?>> Dicts(object? v) => ((List<object?>)v!).Cast<Dictionary<string, object?>>().ToList();
        public static List<object?> Arr(object? v) => (List<object?>)v!;
    }

    public class EmaGoldenTests
    {
        [Fact]
        public void Ema()
        {
            foreach (var c in Fx.List("ema"))
            {
                var e = new Ema(c.D("tau"));
                var steps = Fx.Dicts(c["steps"]);
                for (int i = 0; i < steps.Count; i++)
                {
                    if (i == steps.Count - 1) e.Reset();
                    var v = e.Update(steps[i].D("x"), steps[i].D("t"));
                    Fx.Near(v, steps[i].D("v"), 1e-9, $"tau {c.D("tau")} step {i}");
                }
            }
        }
    }

    public class RuntimeGoldenTests
    {
        [Fact]
        public void Runtime()
        {
            var rt = new RuntimeEstimator();
            // Fixture only records every 7th step (plus the first 7); replay the same plan.
            double t = 0, e = 60;
            var plan = new (string status, double w, int n)[] { ("Discharging", 12, 180), ("Charging", -30, 120), ("Not charging", 0, 12), ("Discharging", 8, 60) };
            var recorded = Fx.List("runtime");
            int ri = 0, i = 0;
            foreach (var (status, w, n) in plan)
            {
                for (int k = 0; k < n; k++)
                {
                    t += 10; e -= w * 10 / 3600;
                    rt.Add(t, e, status);
                    i++;
                    if (i % 7 == 0 || i < 8)
                    {
                        var r = recorded[ri++];
                        Fx.Near(t, r.D("t"));
                        var outp = rt.Estimate(e, r.D("target_wh"), r.D("fallback_w"));
                        var exp = (Dictionary<string, object?>)r["out"]!;
                        foreach (var kv in exp)
                            Fx.Near(Snapshot.Double(outp[kv.Key])!.Value, Snapshot.Double(kv.Value)!.Value, 1e-6, $"step {i} key {kv.Key}");
                    }
                }
            }
            Assert.Equal(recorded.Count, ri);
        }
    }

    public class PowerModelGoldenTests
    {
        [Fact]
        public void Breakdown()
        {
            var m = new PowerModel(5);
            var cases = Fx.List("power_model");
            for (int i = 0; i < cases.Count; i++)
            {
                var c = cases[i];
                var inp = (Dictionary<string, object?>)c["in"]!;
                var raw = new RawPower(inp.D("bat_w"), inp.D("psys"), inp.D("package"), inp.D("core"), inp.D("uncore"), inp.D("dram"));
                var outp = m.Update(raw, inp.B("on_ac"), c.D("t"), inp.D("emission"), inp.B("display_on"));
                Fx.Near(m.Raw.Rest, c.D("raw_rest"), 1e-9, $"step {i} rest");
                foreach (var kv in (Dictionary<string, object?>)c["out"]!)
                {
                    if (kv.Value is string s) Assert.True(outp[kv.Key] as string == s, $"step {i} {kv.Key}");
                    else Fx.Near(Snapshot.Double(outp[kv.Key])!.Value, Snapshot.Double(kv.Value)!.Value, 1e-9, $"step {i} {kv.Key}");
                }
            }
        }
    }

    internal sealed class FakeChargeHW : IChargeHardware
    {
        public bool ThresholdsSupported => true;
        public IReadOnlyList<string> Behaviours => new[] { "auto", "inhibit-charge", "force-discharge" };
        public List<string> Writes = new List<string>();
        public int? Saved;
        public void WriteThresholds(int start, int end)
        {
            // Linux writes them in an order dictated by sysfs; the fixture is normalised below.
            Writes.Add($"start:{start}"); Writes.Add($"end:{end}");
        }
        public void WriteBehaviour(string behaviour) => Writes.Add($"beh:{behaviour}");
        public int? LoadLimit() => null;
        public void SaveLimit(int limit) => Saved = limit;
    }

    public class ChargeGoldenTests
    {
        [Fact]
        public void StateMachine()
        {
            var hw = new FakeChargeHW();
            var c = new ChargeStateMachine(hw, 20);
            foreach (var step in Fx.List("charge"))
            {
                var op = step.S("op");
                var args = (Dictionary<string, object?>)step["args"]!;
                switch (op)
                {
                    case "startup": c.ApplyStartup(); break;
                    case "set_limit": c.SetLimit(args.I("pct")); break;
                    case "start_topup": c.StartTopUp(); break;
                    case "start_discharge": c.StartDischarge(args.I("target")); break;
                    case "cancel": c.Cancel(); break;
                    case "shutdown": c.Shutdown(); break;
                    case "tick": c.Tick(args.I("pct"), args.S("status")); break;
                    default: Assert.Fail($"unknown op {op}"); break;
                }
                var st = (Dictionary<string, object?>)step["state"]!;
                Assert.True(c.Mode.Raw() == st.S("charge_mode"), op);
                Assert.True(c.Limit == st.I("charge_limit"), op);
                Assert.True(c.Target == st.I("charge_target"), op);
                // Compare the set of writes ignoring the sysfs-specific threshold ordering.
                var expected = Fx.Arr(step["writes"]).Select(w => { var a = Fx.Arr(w); return $"{a[0]}:{FmtVal(a[1])}"; }).OrderBy(x => x, StringComparer.Ordinal).ToList();
                var got = hw.Writes.OrderBy(x => x, StringComparer.Ordinal).ToList();
                Assert.True(expected.SequenceEqual(got), $"{op}: expected [{string.Join(",", expected)}] got [{string.Join(",", got)}]");
                hw.Writes.Clear();
            }
        }

        private static string FmtVal(object? v) => v is double d ? ((long)d).ToString() : (v?.ToString() ?? "");
    }

    internal sealed class FakeBrightness : IBrightnessControl
    {
        public bool Available => true;
        public int Max => 1000;
        public int Value = 700;
        public List<int> Sets = new List<int>();
        public int? ReadRaw() => Value;
        public void SetRaw(int v) { Value = v; Sets.Add(v); }
    }

    public class DisplayCalGoldenTests
    {
        [Fact]
        public void SweepAndInterpolation()
        {
            var fx = Fx.Obj("display_cal");
            var bl = new FakeBrightness();
            var d = new DisplayCalibration(bl, null);
            var now = 100.0;
            d.Start(now, true);
            Assert.Equal("running", d.State);
            int guard = 0;
            while (d.State == "running" && guard < 10000)
            {
                now += 0.5;
                var lvl = bl.Value / 1000.0;
                var noise = new[] { 0.3, -0.2, 0.1, 0.0, -0.1 }[(int)(now * 2) % 5];
                d.Tick(3.0 + 6.0 * lvl + noise, now);
                guard++;
            }
            Assert.Equal(fx.S("final_state"), d.State);
            Assert.Equal(Fx.Arr(fx["sets"]).Select(v => (int)(double)v!).ToList(), bl.Sets);
            var table = Fx.Arr(fx["table"]).Select(Fx.Arr).ToList();
            Assert.Equal(table.Count, d.Table.Count);
            for (int i = 0; i < table.Count; i++)
            {
                Assert.Equal((double)table[i][0]!, (double)d.Table[i].raw);
                Fx.Near(d.Table[i].w, (double)table[i][1]!, 1e-9);
            }
            Fx.Near(d.Rest0, fx.D("rest0"), 1e-9);
            foreach (var c in Fx.Dicts(fx["interp"]))
            {
                var raw = c.I("raw");
                var apl = c.D("apl");
                if (c.ContainsKey("w"))
                {
                    if (apl >= 0) d.SetContent(apl, now); else d.SetContent(-1, now - 100);
                    Fx.Near(d.EmissionW(raw, now), c.D("w"), 1e-9, $"raw {raw} apl {apl}");
                }
                else if (c.ContainsKey("w_stale"))
                {
                    d.SetContent(apl, now);
                    Fx.Near(d.EmissionW(raw, now + 61), c.D("w_stale"), 1e-9, "stale");
                }
            }
        }
    }

    public class HistoryGoldenTests
    {
        [Fact]
        public void Downsampling()
        {
            var fx = Fx.Obj("history");
            var h = new History(120, 10);
            foreach (var s in Fx.Dicts(fx["snaps"])) h.Add(s);
            var get = (Dictionary<string, object?>)fx["get"]!;
            foreach (var f in new[] { "sys_w", "soc_w", "bat_pct" })
            {
                var exp = Fx.Arr(get[f]).Select(Fx.Arr).ToList();
                var got = h.Get(f, 60);
                Assert.True(got.Count == exp.Count, f);
                for (int i = 0; i < exp.Count; i++)
                {
                    Fx.Near(got[i].t, (double)exp[i][0]!);
                    Fx.Near(got[i].v, (double)exp[i][1]!, 1e-9, f);
                }
            }
            var all = Fx.Arr(get["all_sys"]);
            Assert.Equal(all.Count, h.Get("sys_w", 1e9).Count);
        }
    }

    public class I18nTests
    {
        [Fact]
        public void ResolveAndFormat()
        {
            Assert.Equal("zh_CN", I18n.ResolveLanguage("system", new[] { "zh-Hans-CN", "en" }));
            Assert.Equal("en", I18n.ResolveLanguage("system", new[] { "en-US" }));
            Assert.Equal("zh_CN", I18n.ResolveLanguage("zh-cn", new[] { "en" }));
            I18n.SetLanguage("en", new string[0]);
            Assert.Equal("Limit 80%", I18n.T("limit", "n", 80));
            Assert.Equal("2 h 5 m", I18n.FmtDuration(125));
            Assert.Equal("12 min", I18n.FmtDuration(12.4));
            I18n.SetLanguage("zh-cn", new string[0]);
            Assert.Equal("2 小时 5 分", I18n.FmtDuration(125));
            Assert.Equal("missingKey", I18n.T("missingKey"));
            var en = new HashSet<string>(I18n.Strings["en"].Keys);
            var zh = new HashSet<string>(I18n.Strings["zh_CN"].Keys);
            Assert.True(en.SetEquals(zh), "zh/en key sets differ: " + string.Join(",", en.Except(zh).Concat(zh.Except(en))));
        }
    }

    public class ProcessBudgetTests
    {
        [Fact]
        public void BudgetDistribution()
        {
            var pb = new ProcessBudget(3, 600);
            pb.Sample(0, 4, () => new (string, double)[0]);
            var top = pb.Sample(10, 10, () => new[] { ("a", 50.0), ("b", 25.0), ("a", 25.0), ("c", 0.0), ("d", 5.0) });
            // floor is min over window = 4, budget 6 W; a=75 of 105
            Assert.Equal(3, top.Count);
            Assert.Equal("a", top[0].name); Fx.Near(top[0].w, 6 * 75.0 / 105.0, 1e-9);
            Assert.Equal("b", top[1].name); Assert.Equal("d", top[2].name);
            // Called again within interval -> cached
            Assert.Equal("a", pb.Sample(11, 20, () => new[] { ("z", 1.0) }).First().name);
        }
    }

    public class JsonTests
    {
        [Fact]
        public void RoundTrip()
        {
            var snap = new Dictionary<string, object?> { ["sys_w"] = 12.5, ["bat_pct"] = 80, ["on_ac"] = true, ["name"] = "屏幕 \"x\"", ["nan"] = double.NaN, ["list"] = new List<object?> { 1.0, "a" } };
            var text = Snapshot.Json(snap, pretty: false);
            var back = Json.ParseObject(text);
            Assert.Equal(12.5, back.D("sys_w"));
            Assert.Equal(80, back.I("bat_pct"));
            Assert.True(back.B("on_ac"));
            Assert.Equal("屏幕 \"x\"", back.S("name"));
            Assert.Equal(-1.0, back.D("nan"));
            Assert.Equal(2, Fx.Arr(back["list"]).Count);
        }
    }
}
